crates/sui-bytecode-src-verifier/src/lib.rs

```diff
@@ -0,0 +1,253 @@
+
+                for (oc_name, on_chain_bytes) in &on_chain_package.module_map {
+                    let on_chain_module_symbol = Symbol::from(oc_name.as_str());
+                    let local_bytes = match local_pkg_bytes.get(&on_chain_module_symbol) {
+                        Some(bytes) => bytes,
+                        None => {
+                            return Err(DependencyVerificationError::LocalDependencyNotFound(
+                                pkg_symbol,
+                                Some(on_chain_module_symbol),
+                            ))
+                        }
+                    };
+
+                    // compare local bytecode to on-chain bytecode to ensure integrity of our dependencies
+                    if local_bytes != on_chain_bytes {
+                        return Err(DependencyVerificationError::ModuleBytecodeMismatch(
+                            pkg_symbol.to_string(),
+                            on_chain_module_symbol.to_string(),
+                            addr,
+                            local_bytes.clone(),
+                            on_chain_bytes.clone(),
+                        ));
+                    }
+
+                    if self.verbose {
+                        println!(
+                            "{}::{} - {} bytes, code matches",
+                            pkg_symbol,
+                            oc_name,
+                            on_chain_bytes.len()
+                        );
+                    }
+                }
+
+                on_chain_module_count += on_chain_package.module_map.len();
+
+                verified_dependencies.insert(
+                    addr,
+                    Dependency {
+                        symbol: symbol.to_string(),
+                        module_bytes: on_chain_package.module_map.clone(),
+                    },
+                );
+            }
+        }
+
+        // total number of modules in packages must match, in addition to each individual module matching
+        if compiled_package.deps_compiled_units.len() != on_chain_module_count {
+            let len = compiled_package.deps_compiled_units.len();
+            return Err(DependencyVerificationError::ModuleCountMismatch(
+                len,
+                on_chain_module_count,
+            ));
+        }
+
+        Ok(DependencyVerificationResult {
+            verified_dependencies,
+        })
+    }
+
+    fn get_module_bytes_map(
+        compiled_package: &CompiledPackage,
+    ) -> HashMap<Symbol, HashMap<Symbol, Vec<u8>>> {
+        let mut map: HashMap<Symbol, HashMap<Symbol, Vec<u8>>> = HashMap::new();
+        compiled_package
+            .deps_compiled_units
+            .iter()
+            .for_each(|(symbol, unit_src)| {
+                let name = unit_src.unit.name();
+                // in the future, this probably needs to specify the compiler version instead of None
+                let bytes = unit_src.unit.serialize(None);
+
```

Comment:
It's subtle, but here, you should be able to extract the module address via 
```
if let CompiledUnitEnum::Module(m) = unit_src {
  let module_addr: AccountAddress = m.address.into_inner();
  // here or later, use module_addr to look up the corresponding module in Sui
} else { unreachable!() }
```

I think that's the missing ingredient that should allow you to kill the use of the `ResolutionGraph`

@tnowacki and @amnn, just noticing how hard it is to work with the internals of the package system due to the script/module bifurcation...

Reply:
@sblackshear I could do this, but a lot of my loop logic is already based around how the resolution graph is structured, and i'd have to rethink it a bit. 

I don't see a strong benefit to cutting out the resolution graph, so given the effort i'd like to leave it as-is

Reply:
I think your code will be much shorter if you get rid of the resolution graph--basically, you can skip straight to line 124 and delete everything above.

It's not just about shorter code, though--the purpose of source verification is to use the package-to-be-published as the ground truth. Here, we have two sources of ground truth: the package itself, and the resolution graph. This raises lots of questions--what happens the `path` passed into `verify_deployed_dependencies` is not the same as the one used to build the `CompiledPackage`? What if the deps in the `ResolutionGraph` don't match the `CompiledPackage`?

By using one source of truth, we don't have to answer these questions.

Reply:
For sensitive logic like this, IMO it's very important to make everything as simple and direct as possible.

Reply:
I did this refactor but i actually think i made the logic worse.  Having addresses associated with modules instead of packages is strange and leads to extra code